File: flink-learning-scala/src/main/scala/com/zhangdi/flink/source/SensorSourceFromRandom.scala
package com.zhangdi.flink.source

import java.util.Calendar

import com.zhangdi.flink.model.SensorReading
import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, SourceFunction}

import scala.util.Random

/**
 * 自定义数据源，返回的数据范型为 {@link SensorReading}
 *
 * 无限生成随机的SensorReading数据。
 *
 * @author di.zhang
 * @date 2020/8/20
 * @time 18:20
 **/
class SensorSourceFromRandom extends RichParallelSourceFunction[SensorReading] {
  //表示数据源是否正常运行。
  var running: Boolean = true

  /**
   * 上下文参数用来发出数据
   *
   * @param ctx
   */
  override def run(ctx: SourceFunction.SourceContext[SensorReading]): Unit = {
    val rand = new Random()


    //使用高斯噪声产生随机温度值
    var curFTemp = (1 to 10).map(i => ("sensor_" + i, rand.nextGaussian() * 20))

    //产生无限数据流
    while (running) {
      curFTemp = curFTemp.map(t => (t._1, t._2 + (rand.nextGaussian() * 0.5)))
      //产生ms为单位的时间戳
      val curTime = Calendar.getInstance.getTimeInMillis
      //发送数据
      curFTemp.foreach(t => ctx.collect(SensorReading(t._1, curTime, t._2)))

      //每间隔1秒发送一条传感器数据
      Thread.sleep(1000)
    }
  }

  /**
   * 定义当取消flink任务时，需要关闭数据源
   */
  override def cancel(): Unit = running = false
}

File: flink-learning-java/src/main/java/com/zhangdi/flink/java/apitest/source/SensorSourceFromRandom.java
package com.zhangdi.flink.java.apitest.source;

import com.zhangdi.flink.java.apitest.model.SensorReading;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;

/**
 * @author zhangdi
 * @description: 随机生成传感器数据
 * @date 2021/1/6 下午5:45
 * @since v1.0
 **/
public class SensorSourceFromRandom extends RichParallelSourceFunction<SensorReading> {

  private volatile boolean running = true;
  private Random rand = null;

  @Override
  public void open(Configuration parameters) throws Exception {
    super.open(parameters);
    rand = new Random();
  }

  @Override
  public void run(SourceContext<SensorReading> ctx) throws Exception {
    List<Tuple2> curFTemp = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).stream()
        .map(i -> new Tuple2<String, Double>("sensor_" + i, rand.nextGaussian() * 20))
        .collect(Collectors.toList());

    while (running) {
      for (Tuple2<String, Double> tuple2 : curFTemp) {
        SensorReading sensorReading = new SensorReading(tuple2.f0, System.currentTimeMillis(),
            tuple2.f1 + (rand.nextGaussian() * 0.5d));
        ctx.collect(sensorReading);
      }
      TimeUnit.SECONDS.sleep(10);
    }
  }

  @Override
  public void cancel() {
    running = false;
  }
}
